### PR TITLE
Nicolas/quickfixes

### LINKED ISSUE #9298: Bug "Save preferences"
**Context:**
Claire is using the feature of "Save" & "Load" Research Preferences to simplify her search. 
When Claire is trying to save the research with the following parameters: World, Boats, 1 an, Non Exclusif this search is saved, but can't be loaded after. 

Nevertheless, when we put Content Type (Movie), Genre (Comedy), Country of Origin (France) this kind of search is loaded with no problems. 

Maybe it's the avails filter that induces the problem?

### LINKED ISSUE #9309: Territory Group Order
Following the improvements of the Country and Territories filter, Claire asked us if the country-groups might be ordered  from A-Z (Africa, Asia, Europe, etc.)

### PR DESCRIPTION
**Pull request by n-germain-work**
### fix #9298 
All filter parameters are written in the url. The url having a length limit, too many parameters led to the issue.
Selecting `world` in avails filter was trying to write all 239 countries. Same problem could have happeened with origin countries filter.
To try and prevent this, both territory from avail filter and origin countries now have the following encoding in the url :
- continent name if all countries of a continent are selected
- 2 letters ISOA2 code instead of the full country name (ex: FR for France)

Additionally, `territories`, `territoriesISOA2` and `territoriesISOA3` have been harmonized.

### fix #9309 
`TerritoryGroups` reordered alphabatically.

### CODE REVIEW

**Comment by n-germain-work** [comment]
Seems there is another issue with the spinner :
image

File: libs/model/src/lib/static/groups.ts [hunk 1]
```diff
@@ -216,7 +145,6 @@ export const territoriesGroup: StaticGroup<'territories'>[] = [
       "grenada",
       "haiti",
       "jamaica",
-      "martinique",
```
Comment (phpgeekfr) [inline libs/model/src/lib/static/groups.ts:219]: keep it if this does not create a bug

File: libs/model/src/lib/static/groups.ts [hunk 2]
```diff
@@ -232,36 +160,104 @@ export const territoriesGroup: StaticGroup<'territories'>[] = [
...
-      "australia",
-      "cook-islands",
-      "fiji",
-      "french-polynesia",
-      "guam",
-      "heard-island-and-mcdonald-islands",
-      "kiribati",
-      "marshall",
-      "micronesia",
-      "midway-islands",
```
Comment (phpgeekfr) [inline libs/model/src/lib/static/groups.ts:248]: same

File: libs/contract/src/lib/avails/tests/holdbacks.spec.ts [hunk 3]
```diff
@@ -22,7 +22,7 @@ const nonCollidingTerritoriesB: Territory[] = [
   'lesotho', 'liberia', 'libya', 'liechtenstein',
   'lithuania', 'luxembourg', 'macao', 'madagascar',
   'malawi', 'malaysia', 'maldives', 'mali', 'malta',
-  'marshall', 'martinique', 'mauritania', 'mauritius',
+  'marshall', 'mauritania', 'mauritius',
```
Comment (phpgeekfr) [inline libs/contract/src/lib/avails/tests/holdbacks.spec.ts:25]: same

File: apps/catalog/catalog-e2e/src/fixtures/dashboard/create-avails.ts [hunk 4]
```diff
@@ -75,9 +75,9 @@ export const expectedContract: Partial<Contract> = {
 
 export const expectedTerm = {
   territories: [
-    //all territories in Europe and Latin America, plus Nepal
-    territoriesGroup.map(group => (group.label === 'Europe' || group.label === 'Latin America') && group.items).filter(Boolean),
+    // Nepal plus all territories in Europe and Latin America 
     'nepal',
+    territoriesGroup.map(group => (group.label === 'Europe' || group.label === 'Latin America') && group.items).filter(Boolean)
```
Comment (phpgeekfr) [inline apps/catalog/catalog-e2e/src/fixtures/dashboard/create-avails.ts:80]: @n-germain-work order have changed. As Nepal is part of Asia, I guess this is why it now comes first

File: apps/catalog/catalog-e2e/src/integration/marketplace/deal-buyer-negotiatiation.cy.ts [hunk 5]
```diff
@@ -152,7 +152,7 @@ describe('Deal negotiation', () => {
     assertTableRowData(0, [
       dateToMMDDYYYY(add(seller.term.duration.from, { days: 15 })),
       dateToMMDDYYYY(sub(seller.term.duration.to, { days: 15 })),
-      'Latin America, CIS, Europe',
+      'Latin America, Europe, CIS',
```
Comment (phpgeekfr) [inline apps/catalog/catalog-e2e/src/integration/marketplace/deal-buyer-negotiatiation.cy.ts:155]: @n-germain-work wierd, must be `chrome` that have a different behavior for order :
- firefox : 'CIS, Europe, Latin America'
- chome: 'Latin America, Europe, CIS'

File: apps/catalog/catalog-e2e/src/integration/marketplace/search-title.cy.ts [hunk 6]
```diff
@@ -243,10 +243,9 @@ describe('Movie search in marketplace', () => {
     escapeKey();
     get('origin-countries').should('contain', 'Europe');
     get('save-filter').click();
-    // url should include all european countries, and not contain the group name (Europe)
-    europeanCountries.forEach(country => cy.url().should('include', country));
-    cy.url().should('not.include', 'Europe');
-    cy.url().should('not.include', 'europe');
+    // url should include Europe and not countries of Europe
+    europeanCountries.forEach(country => cy.url().should('not.include', country));
+    cy.url().should('include', 'Europe');
```
Comment (phpgeekfr) [inline apps/catalog/catalog-e2e/src/integration/marketplace/search-title.cy.ts:248]: @n-germain-work we now check that we have only `Europe` in URL

File: apps/financiers/financiers/src/app/marketplace/title/list/list.component.html [hunk 7]
```diff
@@ -39,9 +39,16 @@ <h1 class="mat-display-2 page-title">All Projects</h1>
     </ng-template>
 
     <ng-template filter label="Country of Origin" [form]="searchForm.originCountries">
-      <chips-autocomplete [form]="searchForm.originCountries" placeholder="Country of Origin" scope="territories">
-        <mat-label>Type and select Country</mat-label>
-      </chips-autocomplete>
+      <mat-form-field appearance="outline">
```
Comment (phpgeekfr) [inline apps/financiers/financiers/src/app/marketplace/title/list/list.component.html:42]: @n-germain-work old component was broken on financiers app. I added the new one like other apps

File: apps/financiers/financiers/src/app/marketplace/title/list/list.component.ts [hunk 8]
```diff
@@ -64,7 +64,7 @@ export class ListComponent implements OnInit, OnDestroy, AfterViewInit {
           const currentSearch = JSON.stringify(search);
           if (this.previousSearch !== currentSearch && this.searchForm.page.value !== 0) {
             this.searchForm.page.setValue(0, { onlySelf: false, emitEvent: false });
-            encodeUrl<MovieAvailsSearch>(this.router, this.route, { search: this.searchForm.value });
+            encodeAvailsSearchUrl(this.router, this.route, { search: this.searchForm.value });
```
Comment (phpgeekfr) [inline apps/financiers/financiers/src/app/marketplace/title/list/list.component.ts:67]: @n-germain-work you forgot to update code for this app (originCountries)

File: libs/model/src/lib/algolia.ts [hunk 9]
```diff
@@ -124,7 +124,7 @@ export interface MovieSearch extends AlgoliaSearch {
 
 export interface MovieAvailsSearch {
   search?: MovieSearch;
-  avails?: AvailsFilter | CalendarAvailsFilter | MapAvailsFilter;
+  avails?: AvailsFilter;
```
Comment (phpgeekfr) [inline libs/model/src/lib/algolia.ts:127]: @n-germain-work for a better typing, `avails` can now only be `AvailsFilter`. I created `MovieMixedSearch` for analytics purpose where `avails` can be `AvailsFilter | CalendarAvailsFilter | MapAvailsFilter`

File: libs/model/src/lib/static/groups.ts [hunk 10]
```diff
@@ -30,77 +30,6 @@ export const mediaGroup: StaticGroup<'medias'>[] = [{
 }];
 
 export const territoriesGroup: StaticGroup<'territories'>[] = [
```
Comment (phpgeekfr) [inline libs/model/src/lib/static/groups.ts:32]: @n-germain-work I added back "Martinique" and the other one

File: libs/model/src/lib/static/types.ts [hunk 11]
```diff
@@ -157,6 +158,8 @@ export type UnitBox = keyof typeof unitBox;
 
 export type Territory = keyof typeof territories;
 export type TerritoryValue = typeof territories[Territory];
+type TerritoryISOA2 = keyof typeof territoriesISOA2;
```
Comment (phpgeekfr) [inline libs/model/src/lib/static/types.ts:161]: no need for export

File: libs/utils/src/lib/form/form-state-url-encoder.ts [hunk 12]
```diff
@@ -30,3 +40,77 @@ export function encodeUrl<T>(
     });
   }
 }
+
+type EncodedTerritory = Territory | TerritoryISOA2Value | TerritoryGroup;
+interface MovieAvailsSearchWithEncodedTerritories {
+  search?: Omit<MovieSearch, 'originCountries'> & { originCountries: EncodedTerritory[] };
+  avails?: Omit<AvailsFilter, 'territories'> & { territories: EncodedTerritory[] };
+}
```
Comment (phpgeekfr) [inline libs/utils/src/lib/form/form-state-url-encoder.ts:48]: @n-germain-work EncodedTerritory and specific `MovieAvailsSearch` is now only used in this file for encode & decode purposes

File: libs/utils/src/lib/form/form-state-url-encoder.ts [hunk 13]
```diff
@@ -30,3 +40,77 @@ export function encodeUrl<T>(
...
+
+export function decodeAvailsSearchUrl(route: ActivatedRoute) {
+  const data = decodeUrl<MovieAvailsSearchWithEncodedTerritories>(route);
+  const search: MovieAvailsSearch = {
+    search: { ...data.search, originCountries: decodeTerritories(data.search?.originCountries) },
+  };
+  if (data.avails) search.avails = { ...data.avails, territories: decodeTerritories(data.avails?.territories) };
+  return search;
+}
```
Comment (phpgeekfr) [inline libs/utils/src/lib/form/form-state-url-encoder.ts:65]: @n-germain-work I refactored with new typing, allowing less `as Territory`

File: libs/utils/src/lib/form/form-state-url-encoder.ts [hunk 14]
```diff
@@ -30,3 +40,77 @@ export function encodeUrl<T>(
...
+  const continents: TerritoryGroup[] = [];
+  const countriesISOA2: TerritoryISOA2Value[] = [];
+  for (const continent of territoriesGroup) {
+    if (continent.items.every(country => countries.includes(country))) {
+      continents.push(continent.label as TerritoryGroup);
+      countries = countries.filter(country => !continentCountries(continent.label as TerritoryGroup).includes(country));
+    }
+  }
+  for (const country of countries) {
+    if (territoriesISOA2[country]) {
```
Comment (phpgeekfr) [inline libs/utils/src/lib/form/form-state-url-encoder.ts:78]: @n-germain-work  this allow to keep countries that does not have a ISOA2 code

**Review by phpgeekfr — APPROVED** [review]

**Pull request merged**